x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,440 @@
+				if err != nil {
+					p.logger.Error(errors.Wrap(err, "bufferedIORead failed"))
+					return
+				}
+
+				offset := 0
+				for {
+					log, err := reader.ReadString('\n')
+					if log == "" {
+						break
+					}
+
+					if err != nil {
+						if err == io.EOF {
+							// create event for last line
+							offset += len([]byte(log))
+							err = p.forwardEvent(createEvent(log, offset, s3Info))
+							if err != nil {
+								p.logger.Error(errors.Wrap(err, "forwardEvent failed"))
+							}
+							break
+						}
+
+						p.logger.Error(errors.Wrap(err, "ReadString failed"))
+						break
+					}
+
+					// create event per log line
+					offset += len([]byte(log))
+					err = p.forwardEvent(createEvent(log, offset, s3Info))
+					if err != nil {
+						p.logger.Error(errors.Wrap(err, "forwardEvent failed"))
+					}
+				}
+			}(s3Infos[i])
+		}
+		wg.Wait()
+	}
+}
+
+func bufferedIORead(svc s3iface.S3API, s3Info s3Info) (*bufio.Reader, error) {
+	s3GetObjectInput := &s3.GetObjectInput{
+		Bucket: awssdk.String(s3Info.name),
+		Key:    awssdk.String(s3Info.key),
+	}
+	req := svc.GetObjectRequest(s3GetObjectInput)
+
+	resp, err := req.Send()
+	if err != nil {
+		return nil, errors.Wrap(err, "s3 get object request failed")
+	}
+
+	return bufio.NewReader(resp.Body), nil
+}
+
+func (p *Input) forwardEvent(event *beat.Event) error {
+	forwarder := harvester.NewForwarder(p.outlet)
+	d := &util.Data{Event: *event}
+	err := forwarder.Send(d)
+	if err != nil {
+		return errors.Wrap(err, "forwarder send failed")
+	}
+	return nil
+}
+
+func deleteMessage(queueURL string, messagesReceiptHandle string, svcSQS *sqs.SQS) error {
+	deleteMessageInput := &sqs.DeleteMessageInput{
+		QueueUrl:      awssdk.String(queueURL),
+		ReceiptHandle: awssdk.String(messagesReceiptHandle),
+	}
+
+	req := svcSQS.DeleteMessageRequest(deleteMessageInput)
+	_, err := req.Send()
+	if err != nil {
+		return errors.Wrap(err, "DeleteMessageRequest failed")
+	}
+	return nil
+}
+
+func createEvent(log string, offset int, s3Info s3Info) *beat.Event {
+	f := common.MapStr{
+		"message": log,
+		"log": common.MapStr{
+			"offset":    int64(offset),
+			"file.path": constructObjectURL(s3Info),
+		},
+		"aws": common.MapStr{
+			"s3": common.MapStr{
+				"bucket": common.MapStr{
+					"name": s3Info.name,
+					"arn":  s3Info.arn},
+				"object.key": s3Info.key,
+			},
+		},
+		"cloud": common.MapStr{
+			"provider": "aws",
+			"region":   s3Info.region,
+		},
+	}
+	return &beat.Event{
+		Meta:      common.MapStr{"id": makeEventID(s3Info, offset)},
+		Timestamp: time.Now(),
+		Fields:    f,
+	}
+}
+
+func constructObjectURL(info s3Info) string {
+	return "https://" + info.name + ".s3-" + info.region + ".amazonaws.com/" + info.key
+}
+
+// makeTopicID returns a short sha256 hash of the bucket name + object key name + offset.
+func makeEventID(s3Info s3Info, offset int) string {
+	h := sha256.New()
+	h.Write([]byte(s3Info.name + s3Info.key + "-" + strconv.Itoa(offset)))
```

Comment:
I think it would better to have <hash>-<offset> then the keys are naturally sortable. Plus you only need to compute the hash one time for the whole object.

Reply:
ah. Maybe they aren't easily sortable, but at least they are somewhat rational to a human.

Reply:
I guess they could be sortable if you padded out the offset like `abcdabcdef-0000123` (assuming you know the maximum offset value a priori).

Reply:
According to aws: `Individual Amazon S3 objects can range in size from a minimum of 0 bytes to a maximum of 5 terabytes`. So 5terabytes is the maximum I guess 😬